# Cesar School - Curso FAST

## Desafio 2

Realizar uma análise básica dos dados relacionados às pessoas vacinadas em 2021, na cidade de Recife.  Baixe a planilha “2021 - Perfil das Pessoas Vacinadas” no site Dados Recife no seguinte link:
Link da base de dados:
http://dados.recife.pe.gov.br/dataset/perfil-das-pessoas-vacinadas-covid-19/resource/99b42b09-95af-47de-8411-ab99c380c3ef (Não é este link)

Link correto: http://dados.recife.pe.gov.br/dataset/7ccb3816-0d62-49e1-b39a-3159870883b0/resource/99b42b09-95af-47de-8411-ab99c380c3ef/download/vacinados.csv



### Importando o pandas
### Baixando a base e transformando em DataFrame

In [1]:
import pandas as pd

In [2]:
url = 'http://dados.recife.pe.gov.br/dataset/7ccb3816-0d62-49e1-b39a-3159870883b0/resource/99b42b09-95af-47de-8411-ab99c380c3ef/download/vacinados.csv'

In [3]:
df = pd.read_csv(url, sep=';', encoding='utf-8')

<ipython-input-3-50b8defdf857>:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(url, sep=';', encoding='utf-8')


### Análise Exploratória da Base

In [ ]:
df.head(2)

faixa_etaria idade      sexo raca_cor municipio  \
0  30 a 34 anos    32  FEMININO    PARDA    OLINDA   
1  35 a 39 anos    39  FEMININO    PARDA  IGARASSU   

                             grupo categoria    lote       vacina_fabricante  \
0           TRABALHADORES DA SAÚDE    OUTROS  FF8842  3 - COMIRNATY (PFIZER)   
1  PÚBLICO EM GERAL (18 a 59 anos)    OUTROS  FD7210  3 - COMIRNATY (PFIZER)   

   descricao_dose                                               cnes  \
0             1.0  DS 2: CNES: 6897029 - POLICLÍNICA SALOMÃO KELNER    
1             1.0     DS 6: CNES: 0001392 - MIGUEL DE LIMA VALVERDE.   

   sistema_origem         data_vacinacao  
0             NÃO  2021-10-07 00:00:00.0  
1  Conecta Recife  2021-09-01 00:00:00.0

In [ ]:
df.tail(2)

faixa_etaria idade      sexo raca_cor municipio  \
3041496  15 a 19 anos    18  FEMININO   BRANCA    RECIFE   
3041497  15 a 19 anos    15  FEMININO    PARDA    RECIFE   

                                        grupo categoria    lote  \
3041496       PÚBLICO EM GERAL (18 a 59 anos)    OUTROS  FJ2594   
3041497   PÚBLICO EM GERAL - MENOR DE 18 ANOS    OUTROS  FG3535   

              vacina_fabricante  descricao_dose  \
3041496  3 - COMIRNATY (PFIZER)             2.0   
3041497  3 - COMIRNATY (PFIZER)             2.0   

                                                      cnes sistema_origem  \
3041496    DS 8: CNES: 0001198 - UBT ARISTARCHO DE AZEVEDO            NÃO   
3041497  DS 1: CNES: 000507 - POLICLÍNICA GOUVEIA DE BA...            NÃO   

                data_vacinacao  
3041496  2021-11-22 00:00:00.0  
3041497  2021-11-10 00:00:00.0

In [ ]:
df.shape

(3041498, 13)

In [ ]:
df.columns

Index(['faixa_etaria', 'idade', 'sexo', 'raca_cor', 'municipio', 'grupo',
       'categoria', 'lote', 'vacina_fabricante', 'descricao_dose', 'cnes',
       'sistema_origem', 'data_vacinacao'],
      dtype='object')

In [ ]:
df.dtypes

faixa_etaria          object
idade                 object
sexo                  object
raca_cor              object
municipio             object
grupo                 object
categoria             object
lote                  object
vacina_fabricante     object
descricao_dose       float64
cnes                  object
sistema_origem        object
data_vacinacao        object
dtype: object

In [ ]:
df.describe()

descricao_dose
count    3.041218e+06
mean     1.642971e+00
std      6.778731e-01
min      1.000000e+00
25%      1.000000e+00
50%      2.000000e+00
75%      2.000000e+00
max      4.000000e+00

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3041498 entries, 0 to 3041497
Data columns (total 13 columns):
 #   Column             Dtype  
---  ------             -----  
 0   faixa_etaria       object 
 1   idade              object 
 2   sexo               object 
 3   raca_cor           object 
 4   municipio          object 
 5   grupo              object 
 6   categoria          object 
 7   lote               object 
 8   vacina_fabricante  object 
 9   descricao_dose     float64
 10  cnes               object 
 11  sistema_origem     object 
 12  data_vacinacao     object 
dtypes: float64(1), object(12)
memory usage: 301.7+ MB


In [ ]:
# pd.set_option('display.max_columns', None)
# print(df)

## Responda às seguintes perguntas:

### 1. Calcule e exiba o número total de pessoas vacinadas.
### R. Número total de pessoas vacinadas: 3041498

In [ ]:
total_vacinadas = len(df)
print(f"Número total de pessoas vacinadas: {total_vacinadas}")

Número total de pessoas vacinadas: 3041498


### 2. Qual foi a cidade que tomou mais dose da vacina CORONAVAC - SINOVAC (BUTANTAN) ?
### R. Recife - 2.911.865 de pessoas vacinadas.

In [4]:
cidade_mais_doses_coronavac = df[df['vacina_fabricante'] == '1 - CORONAVAC - SINOVAC (BUTANTAN)']['municipio'].value_counts().idxmax()
print(f"Cidade que tomou mais doses da vacina CORONAVAC: {cidade_mais_doses_coronavac}")

Cidade que tomou mais doses da vacina CORONAVAC: RECIFE


In [10]:
print(df[df['vacina_fabricante'] == '1 - CORONAVAC - SINOVAC (BUTANTAN)']['municipio'].value_counts())

RECIFE                     724100
JABOATÃO DOS GUARARAPES     10022
OLINDA                       9752
PAULISTA                     7864
CAMARAGIBE                   3679
                            ...  
MORADA NOVA                     1
RECANTO DAS EMAS                1
ITAPUÃ DO OESTE                 1
FEIRA DE SANTANA                1
TAMANDARE                       1
Name: municipio, Length: 182, dtype: int64


In [ ]:
contagem_municipio = df['municipio', ].value_counts()
print(contagem_municipio)

RECIFE                        2911865
JABOATÃO DOS GUARARAPES         37913
OLINDA                          33001
PAULISTA                        23319
CAMARAGIBE                      11843
                               ...   
JARDIM BRASIL                       1
ARMAÇÃO DOS BÚZIOS                  1
GOVERNADOR DIX-SEPT ROSADO          1
ACARAPE                             1
ITAMBE                              1
Name: municipio, Length: 447, dtype: int64


### 3. Crie uma nova coluna chamada ‘’tipo_faixa_etaria”, onde deverá conter os seguintes valores:
- Crianças: 3 a 12 anos
- Adolescentes: 13 a 19 anos
- Jovens adultos: 20 a 39 anos
- Adultos de meia-idade: 40 a 59 anos
- Idosos: 60 anos ou mais

### R. Inicialmente é necessário fazer um tratamento na base. Para criação desta nova coluna, iremos trabalhar com o campo "idade", então, devemos verificar se tem linhas idades vazias. Encontramos 6 registros sem a informação de idade, optamos por apaga-la, pois por ser um número muito pequeno, não influi no resultado.  Após isso, constatamos que o campo idade era um "object", portanto, transformamos ele para númerico. Após conclusão dos procedimentos acima, criamos uma função para criar a nova coluna "tipo_faixa_etaria", conforme solicitado nessa questão.

O método isna() mostra os valores nulos com True e os não nulos como False

In [ ]:
df_empty = df.isna()
df_empty

faixa_etaria  idade   sexo  raca_cor  municipio  grupo  categoria  \
0               False  False  False     False      False  False      False   
1               False  False  False     False      False  False      False   
2               False  False  False     False      False  False      False   
3               False  False  False     False      False  False      False   
4               False  False  False     False      False  False      False   
...               ...    ...    ...       ...        ...    ...        ...   
3041493         False  False  False     False      False  False       True   
3041494         False  False  False     False      False  False       True   
3041495         False  False  False     False      False  False       True   
3041496         False  False  False     False      False  False      False   
3041497         False  False  False     False      False  False      False   

          lote  vacina_fabricante  descricao_dose   cnes  sistema_origem  \
0        False              False           False  False           False   
1        False              False           False  False           False   
2        False              False           False  False           False   
3        False              False           False  False           False   
4        False              False           False  False           False   
...        ...                ...             ...    ...             ...   
3041493  False              False           False  False           False   
3041494  False              False           False  False           False   
3041495  False              False           False  False           False   
3041496  False              False           False  False           False   
3041497  False              False           False  False           False   

         data_vacinacao  
0                 False  
1                 False  
2                 False  
3                 False  
4                 False  
...                 ...  
3041493           False  
3041494           False  
3041495           False  
3041496           False  
3041497           False  

[3041498 rows x 13 columns]

Acrescentando o método any() mostra as colunas que tem pelo menos um campo nulo

In [ ]:
df_empty = df.isna().any()
df_empty

faixa_etaria          True
idade                 True
sexo                  True
raca_cor              True
municipio             True
grupo                False
categoria             True
lote                  True
vacina_fabricante     True
descricao_dose        True
cnes                  True
sistema_origem        True
data_vacinacao        True
dtype: bool

Acrescentando o método sum(), pode-se ver quantos campos nulos tem em cada coluna

In [ ]:
df_empty = df.isna().sum()
print(df_empty)

faixa_etaria              74
idade                      6
sexo                     252
raca_cor                 442
municipio               1573
grupo                      0
categoria            1028992
lote                     293
vacina_fabricante        280
descricao_dose           280
cnes                     280
sistema_origem           280
data_vacinacao           280
dtype: int64


O método dropna() elimina todos os campos nulos. Neste caso, eliminamos todos os 6 valores vazios da coluna "idade".  Como 6 é um valor muito pequeno para o tamanho da base, não influi no resultado.

In [ ]:
df = df.dropna(subset=['idade'])
df_subset = df.isna().any()
df_subset

faixa_etaria          True
idade                False
sexo                  True
raca_cor              True
municipio             True
grupo                False
categoria             True
lote                  True
vacina_fabricante     True
descricao_dose        True
cnes                  True
sistema_origem        True
data_vacinacao        True
dtype: bool

In [ ]:
df_subset = df.isna().sum()
print(df_subset)

faixa_etaria              68
idade                      0
sexo                     252
raca_cor                 442
municipio               1572
grupo                      0
categoria            1028989
lote                     293
vacina_fabricante        280
descricao_dose           280
cnes                     280
sistema_origem           280
data_vacinacao           280
dtype: int64


Com a propriedade dtype, verifica-se que a coluna "idade" é object e é necessário transforma-la em numerica

In [ ]:
df['idade'].dtype

dtype('float64')

Transformando a coluna "idade" para numerica

In [ ]:
df['idade'] = pd.to_numeric(df['idade'], errors='coerce', downcast='integer')
df['idade'].head(2)

0    32.0
1    39.0
Name: idade, dtype: float64

Verifica-se agora que a coluna "idade" é numerica

In [ ]:
df['idade'].dtype

dtype('float64')

In [ ]:
def TipoFaixaEtaria(idade):
    if idade >= 60:
        return "Idosos"
    elif idade >= 40:
        return "Adultos de meia idade"
    elif idade >= 20:
        return "Jovens Adultos"
    elif idade >= 13:
        return "Adolescentes"
    elif idade >= 3:
        return "Crianças"
    else:
        return "Indefinido"

df["tipo_faixa_etaria"] = df['idade'].apply(lambda x:  TipoFaixaEtaria(x))
df.head(5)

faixa_etaria  idade       sexo raca_cor municipio  \
0  30 a 34 anos   32.0   FEMININO    PARDA    OLINDA   
1  35 a 39 anos   39.0   FEMININO    PARDA  IGARASSU   
2  40 a 44 anos   40.0   FEMININO    PARDA  IGARASSU   
3  25 a 29 anos   25.0  MASCULINO   BRANCA    RECIFE   
4  60 a 64 anos   61.0   FEMININO   BRANCA    RECIFE   

                             grupo                        categoria    lote  \
0           TRABALHADORES DA SAÚDE                           OUTROS  FF8842   
1  PÚBLICO EM GERAL (18 a 59 anos)                           OUTROS  FD7210   
2  PÚBLICO EM GERAL (18 a 59 anos)                           OUTROS  FH8023   
3              SESAU - BUSCA ATIVA  PÚBLICO EM GERAL - 18 A 59 ANOS  FH4751   
4                           IDOSOS                           OUTRAS  FM3802   

        vacina_fabricante  descricao_dose  \
0  3 - COMIRNATY (PFIZER)             1.0   
1  3 - COMIRNATY (PFIZER)             1.0   
2  3 - COMIRNATY (PFIZER)             2.0   
3  3 - COMIRNATY (PFIZER)             2.0   
4  3 - COMIRNATY (PFIZER)             3.0   

                                                cnes  sistema_origem  \
0  DS 2: CNES: 6897029 - POLICLÍNICA SALOMÃO KELNER              NÃO   
1     DS 6: CNES: 0001392 - MIGUEL DE LIMA VALVERDE.  Conecta Recife   
2    DS 8: CNES: 0001198 - UBT ARISTARCHO DE AZEVEDO             NÃO   
3  DS 2: CNES: 6897029 - POLICLÍNICA SALOMÃO KELNER              NÃO   
4     DS 6: CNES: 0001392 - MIGUEL DE LIMA VALVERDE.  Conecta Recife   

          data_vacinacao      tipo_faixa_etaria  
0  2021-10-07 00:00:00.0         Jovens Adultos  
1  2021-09-01 00:00:00.0         Jovens Adultos  
2  2021-10-31 00:00:00.0  Adultos de meia idade  
3  2021-11-29 00:00:00.0         Jovens Adultos  
4  2021-12-30 00:00:00.0                 Idosos

### 4. Crie uma nova coluna chamada ‘’regiao”, onde deverá conter valores:
- Metropolitana: Recife, Olinda, Jaboatão dos Guararapes, Paulista, Camaragibe, São Lourenço da Mata, Abreu e Lima, Igarassu, Cabo de Santo Agostinho
- Interior: Cidades que não estão inclusas em Metropolitana


In [ ]:
def Regiao(municipio):
    metropolitanas = ['RECIFE', 'JABOATAO DOS GUARARAPES','OLINDA', 'PAULISTA', 'CAMARAGIBE', 'SAO LOURENCO DA MATA', 'ABREU E LIMA', 'IGARASSU', 'CABO DE SANTO AGOSTINHO']

    if municipio in metropolitanas:
        return 'METROPOLITANA'
    else:
        return 'INTERIOR'

df['regiao'] = df['municipio'].apply(lambda x: Regiao(x))
df.head(5)

faixa_etaria  idade       sexo raca_cor municipio  \
0  30 a 34 anos   32.0   FEMININO    PARDA    OLINDA   
1  35 a 39 anos   39.0   FEMININO    PARDA  IGARASSU   
2  40 a 44 anos   40.0   FEMININO    PARDA  IGARASSU   
3  25 a 29 anos   25.0  MASCULINO   BRANCA    RECIFE   
4  60 a 64 anos   61.0   FEMININO   BRANCA    RECIFE   

                             grupo                        categoria    lote  \
0           TRABALHADORES DA SAÚDE                           OUTROS  FF8842   
1  PÚBLICO EM GERAL (18 a 59 anos)                           OUTROS  FD7210   
2  PÚBLICO EM GERAL (18 a 59 anos)                           OUTROS  FH8023   
3              SESAU - BUSCA ATIVA  PÚBLICO EM GERAL - 18 A 59 ANOS  FH4751   
4                           IDOSOS                           OUTRAS  FM3802   

        vacina_fabricante  descricao_dose  \
0  3 - COMIRNATY (PFIZER)             1.0   
1  3 - COMIRNATY (PFIZER)             1.0   
2  3 - COMIRNATY (PFIZER)             2.0   
3  3 - COMIRNATY (PFIZER)             2.0   
4  3 - COMIRNATY (PFIZER)             3.0   

                                                cnes  sistema_origem  \
0  DS 2: CNES: 6897029 - POLICLÍNICA SALOMÃO KELNER              NÃO   
1     DS 6: CNES: 0001392 - MIGUEL DE LIMA VALVERDE.  Conecta Recife   
2    DS 8: CNES: 0001198 - UBT ARISTARCHO DE AZEVEDO             NÃO   
3  DS 2: CNES: 6897029 - POLICLÍNICA SALOMÃO KELNER              NÃO   
4     DS 6: CNES: 0001392 - MIGUEL DE LIMA VALVERDE.  Conecta Recife   

          data_vacinacao      tipo_faixa_etaria         regiao  
0  2021-10-07 00:00:00.0         Jovens Adultos  METROPOLITANA  
1  2021-09-01 00:00:00.0         Jovens Adultos  METROPOLITANA  
2  2021-10-31 00:00:00.0  Adultos de meia idade  METROPOLITANA  
3  2021-11-29 00:00:00.0         Jovens Adultos  METROPOLITANA  
4  2021-12-30 00:00:00.0                 Idosos  METROPOLITANA

### 5. Conte quantas pessoas receberam a vacina para cada cada  ’regiao” e cada ‘’tipo_faixa_etaria”

In [ ]:
df[['regiao', 'tipo_faixa_etaria']].groupby('regiao')['tipo_faixa_etaria'].value_counts()

regiao         tipo_faixa_etaria    
INTERIOR       Jovens Adultos            28448
               Adultos de meia idade     20014
               Idosos                     3118
               Adolescentes               1631
               Crianças                    172
               Indefinido                  140
METROPOLITANA  Jovens Adultos           971277
               Adultos de meia idade    958279
               Idosos                   782602
               Adolescentes             245022
               Crianças                  30785
               Indefinido                    4
Name: tipo_faixa_etaria, dtype: int64

### 6. Determine a idade média das pessoas vacinadas para  cada  ’regiao” e cada ‘’tipo_faixa_etaria”

In [ ]:
df.groupby(['regiao', 'tipo_faixa_etaria'])['idade'].mean().reset_index(name='idade_media')

regiao      tipo_faixa_etaria  idade_media
0        INTERIOR           Adolescentes    16.899448
1        INTERIOR  Adultos de meia idade    47.373788
2        INTERIOR               Crianças    11.994186
3        INTERIOR                 Idosos    71.567672
4        INTERIOR             Indefinido          NaN
5        INTERIOR         Jovens Adultos    30.311867
6   METROPOLITANA           Adolescentes    16.140306
7   METROPOLITANA  Adultos de meia idade    49.186931
8   METROPOLITANA               Crianças    11.997564
9   METROPOLITANA                 Idosos    70.514246
10  METROPOLITANA             Indefinido     0.500000
11  METROPOLITANA         Jovens Adultos    29.862302

### 7. Identifique e exiba o tipo de vacina mais administrado para cada  ’regiao” e cada ‘’tipo_faixa_etaria”

In [ ]:
df.groupby(['regiao'])[['tipo_faixa_etaria', 'vacina_fabricante' ]].value_counts()

regiao         tipo_faixa_etaria      vacina_fabricante                                
INTERIOR       Jovens Adultos         2 - CHADOX1NCOV-19 - OXFORD/ASTRAZENECA (FIOCRUZ)     12919
               Adultos de meia idade  2 - CHADOX1NCOV-19 - OXFORD/ASTRAZENECA (FIOCRUZ)      9533
               Jovens Adultos         3 - COMIRNATY (PFIZER)                                 7651
                                      1 - CORONAVAC - SINOVAC (BUTANTAN)                     7337
               Adultos de meia idade  1 - CORONAVAC - SINOVAC (BUTANTAN)                     5076
                                      3 - COMIRNATY (PFIZER)                                 4717
               Idosos                 1 - CORONAVAC - SINOVAC (BUTANTAN)                     1224
               Adolescentes           3 - COMIRNATY (PFIZER)                                 1132
               Idosos                 3 - COMIRNATY (PFIZER)                                 1063
                              

### 8. Qual é a proporção de homens e mulheres vacinados para  ’regiao” e cada ‘’tipo_faixa_etaria”?

In [ ]:
df.groupby(['regiao', 'tipo_faixa_etaria'])['sexo'].value_counts()

regiao         tipo_faixa_etaria      sexo     
INTERIOR       Adolescentes           FEMININO        957
                                      MASCULINO       673
                                      OUTROS            1
               Adultos de meia idade  FEMININO      11392
                                      MASCULINO      8580
                                      OUTROS           42
               Crianças               MASCULINO        90
                                      FEMININO         82
               Idosos                 FEMININO       1928
                                      MASCULINO      1190
               Indefinido             1               120
                                      2                20
               Jovens Adultos         FEMININO      17983
                                      MASCULINO     10394
                                      OUTROS           56
METROPOLITANA  Adolescentes           FEMININO     126321
                        